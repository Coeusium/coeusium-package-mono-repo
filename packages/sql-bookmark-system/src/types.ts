export type { ConnectionInfo } from '@almaclaine/mysql-utils';

export interface Bookmark {
  id: string;
  hostname: string;
  pathname: string;
  date_added: Date;
}
