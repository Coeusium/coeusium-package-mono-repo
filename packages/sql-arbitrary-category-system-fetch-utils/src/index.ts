export * from './sql-arbitrary-category-system-fetch-utils';
export type { Category, CategorySet } from './types';
