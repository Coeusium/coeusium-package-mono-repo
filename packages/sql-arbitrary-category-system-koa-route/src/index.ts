export * from './sql-arbitrary-category-system-koa-route';
export type { ConnectionInfo, Category, CategorySet } from './types';
