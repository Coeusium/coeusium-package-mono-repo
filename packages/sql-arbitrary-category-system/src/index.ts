export * from './sql-arbitrary-category-system';
export type { ConnectionInfo, Category, CategorySet } from './types';
