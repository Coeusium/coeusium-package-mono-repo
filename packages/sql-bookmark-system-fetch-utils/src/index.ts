export * from './sql-bookmark-system-fetch-utils';
export type { Bookmark } from './types';
