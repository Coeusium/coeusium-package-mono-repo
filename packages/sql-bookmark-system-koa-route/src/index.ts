export * from './sql-bookmark-system-koa-route';
export type { ConnectionInfo, Bookmark } from './types';
