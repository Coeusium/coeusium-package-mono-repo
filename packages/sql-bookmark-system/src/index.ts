export * from './sql-bookmark-system';
export type { ConnectionInfo, Bookmark } from './types';
