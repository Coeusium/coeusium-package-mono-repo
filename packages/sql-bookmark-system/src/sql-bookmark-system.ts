import { ConnectionInfo, Bookmark } from './types';
import {
  execute,
  setupDatabase,
  readSQLTableFiles,
  checkExists,
  getOneOrDefault,
  getListOrDefault,
  readQuery,
} from '@almaclaine/mysql-utils';
import { URL } from 'url';
import { makeId } from '@almaclaine/general-utils';

export async function setupBookmarkSystem(dbInfo: ConnectionInfo) {
  const files = await readSQLTableFiles(__dirname);
  await setupDatabase(dbInfo, 'sql_bookmark_system', files);
}

export async function destroy() {
  execute.destroyConnections();
}

const BOOKMARK_ID_EXISTS_QUERY = readQuery(__dirname, 'bookmarkIdExists.sql');

export async function bookmarkIdExists(dbInfo: ConnectionInfo, id: string) {
  return checkExists(await execute(dbInfo, BOOKMARK_ID_EXISTS_QUERY, [id]));
}

const INSERT_BOOKMARK_QUERY = readQuery(__dirname, 'insertBookmark.sql');

export async function addBookmark(dbInfo: ConnectionInfo, url: string) {
  let id = makeId();
  const { hostname, pathname } = new URL(url);
  while (await bookmarkIdExists(dbInfo, id)) id = makeId();
  await execute(dbInfo, INSERT_BOOKMARK_QUERY, [id, hostname, pathname]);
  return id;
}

const GET_BOOKMARK_BY_ID_QUERY = readQuery(__dirname, 'getBookmark.sql');

export async function getBookmark(dbInfo: ConnectionInfo, id: string) {
  return getOneOrDefault<Bookmark>(
    await execute(dbInfo, GET_BOOKMARK_BY_ID_QUERY, [id]),
    null,
  );
}

const LIST_BOOKMARKS_QUERY = readQuery(__dirname, 'listBookmarks.sql');

export async function listBookmarks(
  dbInfo: ConnectionInfo,
  page = 0,
  limit = 20,
) {
  const offset = `${limit * page}`;
  return getListOrDefault<Bookmark>(
    await execute(dbInfo, LIST_BOOKMARKS_QUERY, [`${limit}`, offset]),
    [],
  );
}

const DELETE_BOOKMARK_QUERY = readQuery(__dirname, 'deleteBookmark.sql');

export async function deleteBookmark(dbInfo: ConnectionInfo, id: string) {
  await execute(dbInfo, DELETE_BOOKMARK_QUERY, [id]);
}

(async () => {
  await setupBookmarkSystem({
    host: 'localhost',
    user: 'root',
    database: 'sql_bookmark_system',
    password: process.env.MYSQL_PW,
  });
})();
