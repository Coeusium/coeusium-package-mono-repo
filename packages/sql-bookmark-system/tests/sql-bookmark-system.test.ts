import * as sqlBookmarkSystem from '../src';

describe('sql-bookmark-system', () => {
  it('Has Correct API', () => {
    const keys = Object.keys(sqlBookmarkSystem);
    expect(keys.length).toBe(7);
    expect(keys).toMatchSnapshot();
  });
});
