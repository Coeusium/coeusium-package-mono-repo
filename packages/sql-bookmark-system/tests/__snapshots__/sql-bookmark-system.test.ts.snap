// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sql-bookmark-system Has Correct API 1`] = `
Array [
  "setupBookmarkSystem",
  "destroy",
  "bookmarkIdExists",
  "addBookmark",
  "getBookmark",
  "listBookmarks",
  "deleteBookmark",
]
`;
